### PR TITLE
Ecoacoustics preprocessing and data downloading

### PR DESCRIPTION
**Pull request by DavidHurst**

### CODE REVIEW

File: bolts/data/datasets/audio/ecoacoustics.py [hunk 1]
```diff
@@ -1 +1,224 @@
...
+                z.extractall(self.base_dir)
+            else:
+                self.log(f"{fname} already downloaded.")
+
+    def _extract_metadata(self) -> None:
+        """Extract information such as labels from relevant csv files, combining them along with
+        information on processed files to produce a master file."""
+        self.log("Extracting metadata.")
+
+        Extension = Literal[".pt", ".wav"]
```
Comment (olliethomas) [inline bolts/data/datasets/audio/ecoacoustics.py:156]: I'd move this to the top of the file with `SoundscapeAttr`

**Comment by olliethomas** [comment]
When I run the test, I get 
```
bolts/data/datasets/audio/ecoacoustics.py:72: in __init__
    self._download_files()
bolts/data/datasets/audio/ecoacoustics.py:147: in _download_files
    z.extractall(self.base_dir)
../../anaconda3/envs/pal-bolts/lib/python3.9/zipfile.py:1633: in extractall
    self._extract_member(zipinfo, path, pwd)
../../anaconda3/envs/pal-bolts/lib/python3.9/zipfile.py:1686: in _extract_member
    with self.open(member, pwd=pwd) as source, \
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <zipfile.ZipFile file=<_io.BytesIO object at 0x7ff379f3e5e0> mode='r'>, name = <ZipInfo filename='EC_BIRD/.DS_Store' compress_type=deflate filemode='-rw-r--r--' external_attr=0x4000 file_size=6148 compress_size=295>, mode = 'r', pwd = None

    def open(self, name, mode="r", pwd=None, *, force_zip64=False):
        """Return file-like object for 'name'.
    
        name is a string for the file name within the ZIP file, or a ZipInfo
        object.
    
        mode should be 'r' to read a file already in the ZIP file, or 'w' to
        write to a file newly added to the archive.
    
        pwd is the password to decrypt files (only used for reading).
    
        When writing, if the file size is not known in advance but may exceed
        2 GiB, pass force_zip64 to use the ZIP64 format, which can handle large
        files.  If the size is known in advance, it is best to pass a ZipInfo
        instance for name, with zinfo.file_size set.
        """
        if mode not in {"r", "w"}:
            raise ValueError('open() requires mode "r" or "w"')
        if pwd and not isinstance(pwd, bytes):
            raise TypeError("pwd: expected bytes, got %s" % type(pwd).__name__)
        if pwd and (mode == "w"):
            raise ValueError("pwd is only supported for reading files")
        if not self.fp:
            raise ValueError(
                "Attempt to use ZIP archive that was already closed")
    
        # Make sure we have an info object
        if isinstance(name, ZipInfo):
            # 'name' is already an info object
            zinfo = name
        elif mode == 'w':
            zinfo = ZipInfo(name)
            zinfo.compress_type = self.compression
            zinfo._compresslevel = self.compresslevel
        else:
            # Get info object for name
            zinfo = self.getinfo(name)
    
        if mode == 'w':
            return self._open_to_write(zinfo, force_zip64=force_zip64)
    
        if self._writing:
            raise ValueError("Can't read from the ZIP file while there "
                    "is an open writing handle on it. "
                    "Close the writing handle before trying to read.")
    
        # Open for reading:
        self._fileRefCnt += 1
        zef_file = _SharedFile(self.fp, zinfo.header_offset,
                               self._fpclose, self._lock, lambda: self._writing)
        try:
            # Skip the file header:
            fheader = zef_file.read(sizeFileHeader)
            if len(fheader) != sizeFileHeader:
                raise BadZipFile("Truncated file header")
            fheader = struct.unpack(structFileHeader, fheader)
            if fheader[_FH_SIGNATURE] != stringFileHeader:
>               raise BadZipFile("Bad magic number for file header")
E               zipfile.BadZipFile: Bad magic number for file header

../../anaconda3/envs/pal-bolts/lib/python3.9/zipfile.py:1523: BadZipFile
=============================================================================================================================================== short test summary info ================================================================================================================================================
FAILED tests/data_test.py::test_audio_dataset[False] - RuntimeError: Indices file not found at location /Users/ot44/Data/Ecoacoustics.Have you downloaded it?
FAILED tests/data_test.py::test_audio_dataset[True] - zipfile.BadZipFile: Bad magic number for file header

```

**Comment by olliethomas** [comment]
i think the test for download=False should be split into 2 tests. 1) The dataset has not been downloaded and download=False is passed in, and 2) the dataset has been downloaded and download=False is passed.

File: bolts/data/datasets/audio/ecoacoustics.py (hunk 1 above)
Comment (DavidHurst) [reply bolts/data/datasets/audio/ecoacoustics.py:156]: Done

**Comment by DavidHurst** [comment]
> i think the test for download=False should be split into 2 tests. 1) The dataset has not been downloaded and download=False is passed in, and 2) the dataset has been downloaded and download=False is passed.

I might just unparameterise download like the other tests do since testing download=False is an unlikely situation

**Comment by DavidHurst** [comment]
> When I run the test, I get
> 
> ```
> bolts/data/datasets/audio/ecoacoustics.py:72: in __init__
>     self._download_files()
> bolts/data/datasets/audio/ecoacoustics.py:147: in _download_files
>     z.extractall(self.base_dir)
> ../../anaconda3/envs/pal-bolts/lib/python3.9/zipfile.py:1633: in extractall
>     self._extract_member(zipinfo, path, pwd)
> ../../anaconda3/envs/pal-bolts/lib/python3.9/zipfile.py:1686: in _extract_member
>     with self.open(member, pwd=pwd) as source, \
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> 
> self = <zipfile.ZipFile file=<_io.BytesIO object at 0x7ff379f3e5e0> mode='r'>, name = <ZipInfo filename='EC_BIRD/.DS_Store' compress_type=deflate filemode='-rw-r--r--' external_attr=0x4000 file_size=6148 compress_size=295>, mode = 'r', pwd = None
> 
>     def open(self, name, mode="r", pwd=None, *, force_zip64=False):
>         """Return file-like object for 'name'.
>     
>         name is a string for the file name within the ZIP file, or a ZipInfo
>         object.
>     
>         mode should be 'r' to read a file already in the ZIP file, or 'w' to
>         write to a file newly added to the archive.
>     
>         pwd is the password to decrypt files (only used for reading).
>     
>         When writing, if the file size is not known in advance but may exceed
>         2 GiB, pass force_zip64 to use the ZIP64 format, which can handle large
>         files.  If the size is known in advance, it is best to pass a ZipInfo
>         instance for name, with zinfo.file_size set.
>         """
>         if mode not in {"r", "w"}:
>             raise ValueError('open() requires mode "r" or "w"')
>         if pwd and not isinstance(pwd, bytes):
>             raise TypeError("pwd: expected bytes, got %s" % type(pwd).__name__)
>         if pwd and (mode == "w"):
>             raise ValueError("pwd is only supported for reading files")
>         if not self.fp:
>             raise ValueError(
>                 "Attempt to use ZIP archive that was already closed")
>     
>         # Make sure we have an info object
>         if isinstance(name, ZipInfo):
>             # 'name' is already an info object
>             zinfo = name
>         elif mode == 'w':
>             zinfo = ZipInfo(name)
>             zinfo.compress_type = self.compression
>             zinfo._compresslevel = self.compresslevel
>         else:
>             # Get info object for name
>             zinfo = self.getinfo(name)
>     
>         if mode == 'w':
>             return self._open_to_write(zinfo, force_zip64=force_zip64)
>     
>         if self._writing:
>             raise ValueError("Can't read from the ZIP file while there "
>                     "is an open writing handle on it. "
>                     "Close the writing handle before trying to read.")
>     
>         # Open for reading:
>         self._fileRefCnt += 1
>         zef_file = _SharedFile(self.fp, zinfo.header_offset,
>                                self._fpclose, self._lock, lambda: self._writing)
>         try:
>             # Skip the file header:
>             fheader = zef_file.read(sizeFileHeader)
>             if len(fheader) != sizeFileHeader:
>                 raise BadZipFile("Truncated file header")
>             fheader = struct.unpack(structFileHeader, fheader)
>             if fheader[_FH_SIGNATURE] != stringFileHeader:
> >               raise BadZipFile("Bad magic number for file header")
> E               zipfile.BadZipFile: Bad magic number for file header
> 
> ../../anaconda3/envs/pal-bolts/lib/python3.9/zipfile.py:1523: BadZipFile
> =============================================================================================================================================== short test summary info ================================================================================================================================================
> FAILED tests/data_test.py::test_audio_dataset[True] - zipfile.BadZipFile: Bad magic number for file header
> ```

Should be fixed now

File: bolts/data/datasets/audio/ecoacoustics.py [hunk 2]
```diff
@@ -1 +1,224 @@
...
+            for i, segment in enumerate(spectrogram_segments):
+                torch.save(segment, f"{str(self._processed_audio_dir)}\\{waveform_filename}_{i}.pt")
+
+    def _segment_spectrogram(self, specgram: Tensor, segment_len: float) -> list[Tensor]:
+        """
+        Takes a spectrogram and segments it into as many segments of segment_len width as possible.
+        """
+        seg_sz = int(specgram.shape[-1] / (self._AUDIO_LEN / segment_len))
+        segment_boundaries = [(i - seg_sz, i) for i in range(seg_sz, specgram.shape[-1], seg_sz)]
+        return [specgram[:, :, start:end] for start, end in segment_boundaries]
```
Comment (MylesBartlett) [inline bolts/data/datasets/audio/ecoacoustics.py:224]:
```suggestion
        segment_boundaries = [(i - seg_sz, i) for i in range(seg_sz, specgram.shape[-1], seg_sz)]
        return [specgram[:, :, start:end] for start, end in segment_boundaries]
```
return specgram.split(split_size_or_sections=seg_sz, dim=-1)

File: bolts/data/datasets/audio/ecoacoustics.py [hunk 3]
```diff
@@ -1 +1,224 @@
...
+import torchaudio
+import torchaudio.functional as F
+import torchaudio.transforms as T
+from typing_extensions import Literal
+
+from bolts.data.datasets.audio.base import PBAudioDataset
+from bolts.data.datasets.utils import AudioTform, FileInfo
+
+__all__ = ["Ecoacoustics"]
+SoundscapeAttr = Literal["habitat", "site"]
```
Comment (MylesBartlett) [inline bolts/data/datasets/audio/ecoacoustics.py:29]: If this is going to be kept for now, would you mind making it into an Enum? The `target_attr` inti argument can then be of type `Union[str, SoundscapeAttr]` with a casting from str to enum  done with `str_to_enum` in the init (see NICO/CelebA for an example)

**Review by MylesBartlett — APPROVED** [review]

File: bolts/data/datasets/audio/ecoacoustics.py [hunk 4]
```diff
@@ -1 +1,224 @@
-# Placeholder for concrete custom dataset for the ecoacoustics project.
+"""Ecoacostics dataset provided A. Eldridge et al.
+    Alice Eldridge, Paola Moscoso, Patrice Guyot, & Mika Peck. (2018).
+    Data for "Sounding out Ecoacoustic Metrics: Avian species richness
+    is predicted by acoustic indices in temperate but not tropical
+    habitats" (Final) [Data set].
+    Zenodo. https://doi.org/10.5281/zenodo.1255218
+"""
```
Comment (MylesBartlett) [inline bolts/data/datasets/audio/ecoacoustics.py:7]:
```suggestion
"""Ecoacostics dataset provided by A. Eldridge et al.
    Alice Eldridge, Paola Moscoso, Patrice Guyot, & Mika Peck. (2018).
    Data for "Sounding out Ecoacoustic Metrics: Avian species richness
    is predicted by acoustic indices in temperate but not tropical
    habitats" (Final) [Data set].
    Zenodo. https://doi.org/10.5281/zenodo.1255218
"""
```

**Comment by olliethomas** [comment]
What's happening here?

**Pull request merged**